test/Microsoft.AspNetCore.Sockets.Common.Tests/Internal/Formatters/BinaryMessageFormatterTests.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Sockets.Internal.Formatters;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Sockets.Tests.Internal.Formatters
+{
+    public partial class BinaryMessageFormatterTests
+    {
+        [Fact]
+        public void WriteMultipleMessages()
+        {
+            var expectedEncoding = new byte[]
+            {
+                /* length: */ 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                    /* type: */ 0x01, // Binary
+                    /* body: <empty> */
+                /* length: */ 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0E,
+                    /* type: */ 0x00, // Text
+                    /* body: */ 0x48, 0x65, 0x6C, 0x6C, 0x6F, 0x2C, 0x0D, 0x0A, 0x57, 0x6F, 0x72, 0x6C, 0x64, 0x21,
+                /* length: */ 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
+                    /* type: */ 0x03, // Close
+                    /* body: */ 0x41,
+                /* length: */ 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x0C,
+                    /* type: */ 0x02, // Error
+                    /* body: */ 0x53, 0x65, 0x72, 0x76, 0x65, 0x72, 0x20, 0x45, 0x72, 0x72, 0x6F, 0x72
+            };
+
+            var messages = new[]
+            {
+                MessageTestUtils.CreateMessage(new byte[0]),
+                MessageTestUtils.CreateMessage("Hello,\r\nWorld!",MessageType.Text),
+                MessageTestUtils.CreateMessage("A", MessageType.Close),
+                MessageTestUtils.CreateMessage("Server Error", MessageType.Error)
+            };
+
+            var output = new ArrayOutput(chunkSize: 8); // Use small chunks to test Advance/Enlarge and partial payload writing
+            foreach (var message in messages)
+            {
+                Assert.True(MessageFormatter.TryWriteMessage(message, output, MessageFormat.Binary));
+            }
+
+            Assert.Equal(expectedEncoding, output.ToArray());
+        }
+
+        [Theory]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01 }, new byte[0])]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x04, 0x01, 0xAB, 0xCD, 0xEF, 0x12 }, new byte[] { 0xAB, 0xCD, 0xEF, 0x12 })]
+        public void WriteBinaryMessage(byte[] encoded, byte[] payload)
+        {
+            var message = MessageTestUtils.CreateMessage(payload);
+            var output = new ArrayOutput(chunkSize: 8); // Use small chunks to test Advance/Enlarge and partial payload writing
```

Comment:
It might be good to add a case where end of the message is on the chunk boundary